Sparse Spanners using Baswana-Sen

gmyr opened the pull request:
As discussed in #1609, I implemented the spanner algorithm of Baswana and Sen (see [1]) while looking into graph sparsification. I cleaned up the code for this pull request. Please let me know if you have any comments or suggestions!

I have a couple of questions:

1. The algorithm is randomized. Is there any specific way I have to set up the random number generator or can I assume that it has already been properly seeded?
2. Since the algorithm is randomized, some parts are not always executed (one line, to be specific: the "continue" in Line 151 of spanner.py). This means that the tests sometimes do not cover this line. Is there anything I should do about this?
3. The unit tests are quite extensive and take somewhere between 3-4 seconds to finish on my machine. Should I use smaller graphs in the tests?

[1] S. Baswana, S. Sen. A Simple and Linear Time Randomized Algorithm for Computing Sparse Spanners in Weighted Graphs. Random Struct. Algorithms 30(4): 532-563 (2007).

dschult commented:
1) 
If your code uses a random number generator then the functuion should accept an argument ```seed``` which should be processed using the decorator ```@py_random_state(<index>)```
or ```@np_random_state(<index>)``` where <index> is the index of the ```seed``` argument.
Use ```py_random_state``` if your code uses Python's ```random``` module. 
Use ```np_random_state``` if your code uses Numpy's random module.

Then in your code when you want a random number, use ```seed``` as the random number generator (RNG).  For example:   ```seed.random()``` or ```seed.gauss()```.

The decorator is imported using  ```from networkx.tools import py_random_state``` or similar.

Note: if your code doesn't create random numbers but calls a function that creates random numbers, then you must still include an input argument for the seed processed the same way as above and then passed into the function that creates the random numbers. This is the only way to make sure that random numbers don't affect the global RNG if the user passes in a ```seed``` parameter.

2) When testing it is probably a good idea to set the seed value. Then it is consistent from one test to the next. Ideally, you would test it with different seed values that as a whole cover all the code. But sometimes that is very difficult to arrange. 

3) I think it would be better to shrink the size of the tests to spped them up unless there is a specific trait of the large graphs that we want to test.

dschult commented:
Also, since this is a new module we need to update the ```/doc/reference``` files so they find and include the function in the docs.  I will try to make this and the above updates if you dont get to it soon. So no worries.    Thanks again for this!

gmyr commented:
@dschult: I changed the code so that it follows your conventions of using RNGs. I also made the unit tests smaller (about 0.1s running time on my machine) and fixed a seed that gives us 100% coverage.

gmyr commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+    return H
+
+
+def _stretch_to_k(stretch):
```
You're right, I changed it accordingly.

gmyr commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+    # clustering is a dictionary that maps nodes in a cluster to the
+    # cluster center
+    clustering = {v: v for v in G.nodes}
+    sample_prob = math.pow(G.number_of_nodes(), - 1.0 / k)
```
Nice! I changed this.

gmyr commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+    # establish unique edge weights, even for unweighted graphs
+    for u, v in G.edges():
+        if not weight:
+            residual_graph[u][v]['weight'] = (id(u), id(v), 0)
```
Exactly, the paper assumes (without loss of generality) that the edge weights are unique. After looking through the python docs for a while, using ```id()``` seemed to be a suitable way to achieve this.

dschult commented:
Awesome.  I also noticed that your module is named the same as your function.  Can you change the name of the module to something different (and slightly more general in case other functions get added)?  This name-change leads to small changes in the rst and __init__ files which are kind of a pain, but we don't clobber the name spanner that way.  Perhaps ```spanning.py```, but maybe there is a better name?

gmyr commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+    return lightest_edge_neighbor, lightest_edge_weight
+
+
+def _closest_sampled_center(lightest_edge_weight, neighboring_sampled_centers):
```
Great suggestion, thanks! Not only does this simplify the code, it also allows me to address the issue you raised above in a different way: For unweighted graphs I can now set the weight in ```_setup_residual_graph``` to ```(id(u), id(v))``` instead of ```(id(u), id(v), 0)```.

gmyr commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+        if not weight:
+            residual_graph[u][v]['weight'] = (id(u), id(v), 0)
+        else:
+            residual_graph[u][v]['weight'] = (G[u][v][weight], id(u), id(v))
```
I changed this a bit, see my comment on your next comment below.

dschult commented on networkx/algorithms/spanner.py:
```diff
@@ -0,0 +1,341 @@
+                for neighbor in residual_graph.adj[v]:
+                    neighbor_cluster = clustering[neighbor]
+                    neighbor_weight = lightest_edge_weight[neighbor_cluster]
+                    if neighbor_cluster is closest_center or neighbor_weight < closest_center_weight:
```
I believe dict's use hash value and then equals when hash values match.
Nodes can be compared using ```==```

gmyr replied:
Ok, I will fix this.

gmyr commented:
Well, on an abstract level a spanner is just a distance-preserving sparsifier of a graph. So it might be fitting to call the module _sparsifier_ or _sparsification_. Maybe someone will add cut-sparsifiers, spectral sparsifiers, distance-preservers (these are distinct from spanners), more general spanners etc. to the module later on. What do you think?

dschult commented:
I like ```sparsifiers.py``` as a general term with plural to hopefully indicate long term multiple approaches.  Thanks!

gmyr commented:
I renamed the module. I also tried to make the required changes in ```/doc/reference``` but could not get the documentation to work properly. @dschult: Could you take care of this?

gmyr commented:
It seems there has been a hickup in the Mac OS tests on travis. Could you restart these?

dschult commented:
Is this ready to merge then?

gmyr commented:
I have nothing more to add at this time, so yes.
